src/stream/callback.rs

```diff
@@ -218,75 +231,184 @@ impl Default for Callback {
 }
 
 impl super::Stream {
+    /// Attempts to downloads the [`Stream`](super::Stream)s resource.
+    /// This will download the video to <video_id>.mp4 in the current working directory.
+    /// Takes an [`Callback`](crate::stream::callback::Callback)
+    #[doc(cfg(feature = "callback"))]
+    #[inline]
+    pub async fn download_callback(&self, callback: Callback) -> Result<PathBuf> {
+        self.wrap_callback(|channel| {
+            self.internal_download(channel)
+        }, callback).await
+    }
+
+    /// Attempts to downloads the [`Stream`](super::Stream)s resource.
+    /// This will download the video to <video_id>.mp4 in the provided directory. 
+    /// Takes an [`Callback`](crate::stream::callback::Callback)
+    #[doc(cfg(feature = "callback"))]
+    #[inline]
+    pub async fn download_to_dir_callback<P: AsRef<Path>>(
+        &self,
+        dir: P,
+        callback: Callback,
+    ) -> Result<PathBuf> {
+        self.wrap_callback(|channel| {
+            self.internal_download_to_dir(dir, channel)
+        }, callback).await
+    }
+
+    /// Attempts to downloads the [`Stream`](super::Stream)s resource.
+    /// This will download the video to the provided file path.
+    /// Takes an [`Callback`](crate::stream::callback::Callback)
+    #[doc(cfg(feature = "callback"))]
+    #[inline]
+    pub async fn download_to_callback<P: AsRef<Path>>(&self, path: P, callback: Callback) -> Result<()> {
+        let _ = self.wrap_callback(|channel| {
+            self.internal_download_to(path, channel)
+        }, callback).await?;
+        Ok(())
+    }
+
+    async fn wrap_callback<F: Future<Output = Result<PathBuf>>>(
+        &self,
+        to_wrap: impl FnOnce(Option<InternalSender>)-> F,
+        mut callback: Callback
+    ) -> Result<PathBuf> {
+        let wrap_fut = to_wrap(Some(callback.internal_sender.clone()));
+        let aid_fut = self.on_progress(
+            callback.internal_receiver.take().expect("Callback cannot be used twice"),
+            std::mem::take(&mut callback.on_progress),
+        );
+        let (result, _) = futures::future::join(wrap_fut, aid_fut).await;
+
+        let path = result.as_ref().map(|p| p.clone()).ok();
+
+        Self::on_complete(std::mem::take(&mut callback.on_complete), path).await;
+
+        result
+    }
+
     #[inline]
-    pub(crate) async fn on_progress(mut receiver: Receiver<usize>, on_progress: OnProgressType) {
-        let counter = Mutex::new(100);
+    async fn on_progress(&self, mut receiver: Receiver<InternalSignal>, on_progress: OnProgressType) {
```

Comment:
This method has way too much code repetition. I would like to refactor especially the slow variants, but didn't came up with a clean/working way of doing so.